Model patcher

jiqing-feng opened the pull request:
This PR enables ipex llama model by patching functions and classes, and it has 30% speed-up than the original optimization. 
The ipex optimization ops will be released soon, and I will add the CI tests once it released. We can focus on the integration for now.

BTW, this PR includes #566, and I will rebase it after #566 is merged.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-intel/pr_567). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

echarlaix commented on optimum/exporters/ipex/llama_functions.py:
```diff
@@ -0,0 +1,362 @@
+) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+    bsz, q_len, _ = hidden_states.size()
+    concat_qkv = None
+    if hasattr(self, "concat_qkv") and self.concat_qkv is not None:
```
where does the `concat_qkv` attribute comes from ?

jiqing-feng replied:
Sorry for the mistake, I haven't integrate `concat_qkv` for now, will remove it.

echarlaix commented on optimum/exporters/ipex/llama_functions.py:
```diff
@@ -0,0 +1,362 @@
+from transformers.modeling_outputs import BaseModelOutputWithPast
+
+
+def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
+    """
+    This is the equivalent of torch.repeat_interleave(x, dim=1, repeats=n_rep). The hidden states go from (batch,
+    num_key_value_heads, seqlen, head_dim) to (batch, num_attention_heads, seqlen, head_dim)
+    """
+    batch, num_key_value_heads, slen, head_dim = hidden_states.shape
+    if n_rep == 1:
+        return hidden_states
+    hidden_states = hidden_states[:, :, None, :, :].expand(batch, num_key_value_heads, n_rep, slen, head_dim)
+    return hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
```
from what I see it doesn't look like there is any difference with https://github.com/huggingface/transformers/blob/593230f0a1150ea9c0477b9d859f25daf73c8c33/src/transformers/models/llama/modeling_llama.py#L223 why redefining it ?

jiqing-feng replied:
Will import this func from transformers.

echarlaix commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -325,10 +348,13 @@ def __init__(
         self.model_dtype = kwargs.get("model_dtype", self.dtype)
+        self._dtype = self.model_dtype
         self.use_cache = "past_key_values" in self.input_names
+        self.is_ipex_exported = kwargs.get("is_ipex_exported", None)
```
what if we load an already exported model : `export=False` -> `_from_transformers` won't be called, `is_ipex_exported` will always be set to None

jiqing-feng replied:
I think we can check the model's arch to decide if `is_ipex_exported` is True or not, WDYT?

jiqing-feng replied:
Sure, will check if the original_name in the graph model is in our supported list or not.

echarlaix commented on optimum/exporters/ipex/model_patcher.py:
```diff
@@ -0,0 +1,71 @@
+    return model
+
+
+def export_model(model):
+    if isinstance(model, LlamaForCausalLM):
+        export_llama_model(model)
+        return True
+
+    return False
```
Could you add it as a ModelPatcher : https://github.com/huggingface/optimum/blob/8b2c81149e267105e1d2cd9bc2346363100869b1/optimum/exporters/onnx/model_patcher.py#L114

jiqing-feng replied:
Hi Ella, thanks for your review. I think we'd better not use unpatch. You can see I replaced some layers here and deleted the original layer, see https://github.com/jiqing-feng/optimum-intel/blob/model_patcher/optimum/exporters/ipex/llama_functions.py#L280-L297. If we use unpatch, then we need to keep both weights of the 2 kinds of layers, and the memory will increase.

Of course, I can do it if you think it is more reasonable, please let me know your decision. Thx!

echarlaix commented on optimum/exporters/ipex/llama_functions.py:
```diff
@@ -0,0 +1,362 @@
+        query = query.view(bsz, q_len, self.num_heads, self.head_dim)
+        key = key.view(bsz, q_len, self.num_key_value_heads, self.head_dim)
+        value = value.view(bsz, q_len, self.num_key_value_heads, self.head_dim)
+        key = self.ipex_rope(
```
could you add comments to explain what all this modifications are needed and  `ipex_rope` / `ipex_scale_dot_product`

jiqing-feng replied:
Sure!

echarlaix commented on optimum/exporters/ipex/llama_functions.py:
```diff
@@ -0,0 +1,362 @@
+    return attn_output, attn_weights, past_key_value
+
+
+def prepare_inputs_for_generation(
```
could you add a comment to explain why this is redefined

jiqing-feng replied:
Yes, will add it in #566

jiqing-feng replied:
Yes

echarlaix commented on optimum/exporters/ipex/__init__.py:
```diff
@@ -0,0 +1 @@
+from .model_patcher import export_model
```
would remove it as we currently don't want user to use at the moment as it will modify the model inplace which can be risky for an uninformed user)

```suggestion
```

jiqing-feng replied:
Sure

echarlaix commented on optimum/exporters/ipex/model_patcher.py:
```diff
@@ -0,0 +1,69 @@
+    return model
+
+
+def export_model(model):
```
I think patch would be more adapted than export here no ? 

also it only patches llama models so would rename it to for now
```suggestion
def _patch_llama_model(model):
```

echarlaix commented on optimum/exporters/ipex/model_patcher.py:
```diff
@@ -0,0 +1,69 @@
+        patch_op(sub_m, target_m, new_op_name, new_op)
+
+
+def export_llama_model(model):
```
```suggestion
def _patch_llama_model(model):
```

echarlaix commented on optimum/exporters/ipex/model_patcher.py:
```diff
@@ -0,0 +1,69 @@
+
+
+def export_llama_model(model):
+    ipex_rope = ApplyRotaryEmbedding(
+        model.config.max_position_embeddings,
+        model.config.hidden_size // model.config.num_attention_heads,
+        model.config.rope_theta,
+        model.config.architectures[0],
+    )
+    ipex_scale_dot_product = IndirectAccessKVCache(text_max_length=model.config.max_position_embeddings)
+    patch_op(model, LlamaAttention, "ipex_rope", ipex_rope)
+    patch_op(model, LlamaAttention, "ipex_scale_dot_product", ipex_scale_dot_product)
+
+    convert_func(model, "prepare_inputs_for_generation", prepare_inputs_for_generation)
+    convert_functions(model, LlamaModel, "forward", llama_model_forward)
+    convert_functions(model, LlamaAttention, "forward", llama_attn_forward)
+    convert_functions(model, LlamaRMSNorm, "forward", llama_layer_norm_forward)
+
+    convert_class(model, LlamaDecoderLayer, _IPEXLlamaDecoderLayerRef, model.config)
```
also because a lot of modifications are applied on the model, could we add couple of tests to verify that the generation behavior for llama models is not unexpectedly impacted ?

jiqing-feng replied:
OK, do you want to check the eager mode ipex model (only use model patcher) or graph model ipex model or both?

jiqing-feng commented on optimum/exporters/ipex/llama_functions.py:
```diff
@@ -0,0 +1,337 @@
+from typing import List, Optional, Tuple, Union
+
+import torch
+from intel_extension_for_pytorch.llm.modules import linear2SiluMul, linearAdd
```
Will add a limitation for ipex version once the ipex 2.3 is released

jiqing-feng commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -116,14 +148,20 @@ def _from_transformers(
-        model = patch_decoder_attention_mask(model)
-        model = ipex.optimize(model, dtype=torch_dtype, level="O1", auto_kernel_selection=True)
-        traced_model = jit_trace(model, task, use_cache)
+        is_ipex_exported = is_ipex_exported_model(model.__class__.__name__)
```
Sure!

jiqing-feng replied:
`model_type` is not a good way because we only optimize `LlamaModelForCausalLM` so just check the class name will be great.

jiqing-feng commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -496,3 +557,28 @@ def _prepare_inputs_for_generation_for_llama(
+    past_key_values: Tuple[Tuple[torch.Tensor]], beam_idx: torch.Tensor
+) -> Tuple[Tuple[torch.Tensor]]:
+
+    if len(past_key_values[0]) == 4 and past_key_values[0][0].shape[-1] == 1:  # discrete kv_cache
+        for layer_past in past_key_values:
+            layer_past[3][layer_past[0].size(-2) - 1] = beam_idx
+        return past_key_values
+    elif len(past_key_values[0]) == 8:
+        for layer_past in past_key_values:
+            layer_past[3][layer_past[0].size(-2) - 1] = beam_idx
+            layer_past[7][layer_past[0].size(-2) - 1] = beam_idx
+        return past_key_values
```
The ipex pkv shape and data layout are different with the original pkv, I will notice it in code comments.

jiqing-feng commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -374,7 +421,21 @@ def _prepare_past_key_values(self, input_ids):
+            past_key_values = tuple(
+                [
+                    (
+                        torch.zeros(1, 0, 0, 1, dtype=torch.long).contiguous(),
+                        torch.zeros([1, 1, 1, 1]).contiguous(),
+                        torch.zeros([1, 1, 1, 1]).contiguous(),
```
It is equal to the prepare pkv for generation for the original model, it is hard to explain due to our ipex pkv data layout. I will try to explain it in code comments.

jiqing-feng commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -374,7 +421,21 @@ def _prepare_past_key_values(self, input_ids):
 
-        if model_type == "bloom":
+        if self.is_ipex_exported:
+            beam_idx_tmp = torch.zeros((2048, input_ids.shape[0]), dtype=torch.long).contiguous()
```
Yes, I will use a parameter like `max_position_embeddings` to replace it.

jiqing-feng commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -374,7 +421,21 @@ def _prepare_past_key_values(self, input_ids):
             num_attention_heads = self.normalized_config.num_attention_heads
 
-        if model_type == "bloom":
+        if self.is_ipex_exported:
```
It is hard code only for llama now, we will add more models like falcon and gptj etc. in the future.

jiqing-feng commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -53,6 +56,34 @@
+IPEX_EXPORTED_LIST = ("LlamaForCausalLM", )
+
+
+def is_ipex_exported_model(model_name):
+    for name in IPEX_EXPORTED_LIST:
+        if model_name == name:
+            return True
+    return False
```
Sure

jiqing-feng commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -53,6 +56,34 @@
+    return False
+
+
+def ipex_jit_trace(model):
```
Yes

jiqing-feng commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -53,6 +56,34 @@
+def ipex_jit_trace(model):
+    sample_inputs = get_dummy_input(model, return_dict=True)
+    model.config.return_dict = False
+    _enable_tpp()
```
_enable_tpp is an ipex specific optimization, will try to explain it in the code comments.

jiqing-feng commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -53,6 +56,34 @@
+    model = ipex.optimize(model.eval(), dtype=model.dtype, inplace=True)
+    trace_model = torch.jit.trace(
+        model,
+        example_kwarg_inputs=sample_inputs,
```
Yes, will add an pytorch version check here.

jiqing-feng commented on optimum/exporters/ipex/model_patcher.py:
```diff
@@ -0,0 +1,69 @@
+from intel_extension_for_pytorch.llm.modules import ApplyRotaryEmbedding, IndirectAccessKVCache
```
Got it.

jiqing-feng commented:
Hi @echarlaix Thanks for your review. I have fixed most of them and still have some issues of the comments and tests, would you please take the next round of review? Thx!

echarlaix commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -304,7 +349,7 @@ def forward(
         if "token_type_ids" in self.input_names:
             inputs["token_type_ids"] = token_type_ids
 
-        outputs = self._call_model(**inputs)
+        outputs = self.model(**inputs)
```
would prefer to keep this changes for https://github.com/huggingface/optimum-intel/pull/566 as we already have the discussion there

echarlaix commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -73,6 +117,7 @@ def __init__(
         self._dtype = self.config.torch_dtype if self.config.torch_dtype is not None else torch.float32
         self.model.to(self._device)
         self.model_save_dir = model_save_dir
+        self.is_ipex_exported = is_model_support_ipex_export(model, self.export_feature)
```
also would rename the attribute to something like `_is_patched_with_ipex`

echarlaix commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -414,7 +481,7 @@ def forward(
             position_ids = attention_mask.long().cumsum(-1) - 1
             position_ids.masked_fill_(attention_mask == 0, 1)
             if past_key_values:
-                position_ids = position_ids[:, -1].unsqueeze(-1)
+                position_ids = position_ids[:, -input_ids.shape[-1] :]
```
why this modification ?

echarlaix replied:
any news on this @jiqing-feng  ?

echarlaix reviewed (commented):
Thanks a lot for iterating on the PR @jiqing-feng

jiqing-feng commented:
Hi @echarlaix . I think I have fixed all your comments, and I also added ipex model generation testing with multi inputs. Would you please help to review it? Thx!

echarlaix commented on tests/ipex/test_modeling.py:
```diff
@@ -255,6 +256,20 @@ def test_pipeline(self, model_arch):
         self.assertEqual(pipe.device, model.device)
         self.assertTrue(all("This is a sample" in item["generated_text"] for item in outputs))
 
+    @parameterized.expand(SUPPORTED_ARCHITECTURES)
```
can you test different beam search / use_cache to True/False and compare the generated outputs with the ones from the original transformers model, also for couple of generation config (beam search but also greedy), this can be done for `IPEX_PATCHED_SUPPORTED_ARCHITECTURES = ("llama",)` as not other architectures currently support the ipex model patching

jiqing-feng commented:
Hi @echarlaix , I have fixed all your comments. Do you mind taking the last round? I think we could merge it first since it does not affect the current version, and will let you know once the new version ipex is released. Thx!

echarlaix commented on tests/ipex/test_modeling.py:
```diff
@@ -255,6 +259,37 @@ def test_pipeline(self, model_arch):
+        generation_config = GenerationConfig(
+            max_new_tokens=16, num_beams=num_beams, do_sample=False, use_cache=use_cache
+        )
+        outputs = model.generate(**tokens, generation_config=generation_config)
+        with torch.no_grad():
+            transformers_outputs = transformers_model(**tokens)
+
+        self.assertIsInstance(outputs.logits, torch.Tensor)
+        # Compare tensor outputs
+        self.assertTrue(torch.allclose(outputs.logits, transformers_outputs.logits, atol=1e-4))
```
this will fail no ? (comparison after `generate` vs one `forward` for the transformers model)
shouldn't it be :
```suggestion
        outputs = model.generate(**tokens, generation_config=generation_config)
        with torch.no_grad():
            transformers_outputs = transformers_model.generate(**tokens, generation_config=generation_config)

        self.assertIsInstance(outputs.logits, torch.Tensor)
        # Compare tensor outputs
        self.assertTrue(torch.allclose(outputs.logits, transformers_outputs.logits, atol=1e-4))
```
It would also make sense to compare both model's output tensor after one forward :
```python
outputs = model(**tokens)
transformers_outputs = transformers_model(**tokens)
```
(can be done for one set of parameters as we don't need to do it for all cases)

echarlaix commented on tests/ipex/test_modeling.py:
```diff
@@ -255,6 +259,37 @@ def test_pipeline(self, model_arch):
+            {
+                "model_arch": IPEX_PATCHED_SUPPORTED_ARCHITECTURES,
+                "use_cache": [True, False],
+                "num_beams": [1, 4],
+                "batch_size": [1, 4],
```
no need to export the model multiple times for different `num_beams` and `batch_size`, I think we can iterate after the export directly

echarlaix commented on tests/ipex/test_modeling.py:
```diff
@@ -255,6 +259,37 @@ def test_pipeline(self, model_arch):
+    def test_ipex_patching(self, test_name, model_arch, use_cache, num_beams, batch_size):
+        model_id = MODEL_NAMES[model_arch]
+        set_seed(SEED)
+        model = IPEXModelForCausalLM.from_pretrained(model_id, export=True)
```
the model will always be exported with `use_cache=True` we should instead set `use_cache` :
```suggestion
        model = IPEXModelForCausalLM.from_pretrained(model_id, export=True, use_cache=use_cache)
```

can you also add a check 

```suggestion
        model = IPEXModelForCausalLM.from_pretrained(model_id, export=True, use_cache=use_cache)
        self.assertTrue(model.use_cache, use_cache)
```

echarlaix commented on tests/ipex/test_modeling.py:
```diff
@@ -255,6 +259,37 @@ def test_pipeline(self, model_arch):
+        tokenizer.pad_token = tokenizer.eos_token
+        texts = ["This is a sample"] * batch_size
+        tokens = tokenizer(texts, padding=True, return_tensors="pt")
+        generation_config = GenerationConfig(
```
could also make sense to add sampling / top-k / top-p ?

echarlaix commented on optimum/exporters/ipex/modeling_utils.py:
```diff
@@ -0,0 +1,317 @@
+    # embed positions
+    hidden_states = inputs_embeds
+
+    if self.gradient_checkpointing and self.training:
+        use_cache = False
```
what is the reason we don't remove this ? 

```suggestion
```

echarlaix commented on optimum/exporters/ipex/modeling_utils.py:
```diff
@@ -0,0 +1,317 @@
+    return attn_output, attn_weights, past_key_value
+
+
+def llama_model_forward(
```
```suggestion
# Adapted from https://github.com/huggingface/transformers/blob/v4.38.2/src/transformers/models/llama/modeling_llama.py#L1130
def _llama_model_forward(
```

echarlaix commented:
also cc @ofirzaf for visibility

jiqing-feng commented:
Hi @echarlaix , thanks for your detailed review! 

I have disabled everything for the next ipex release, you can check on my codes and see that the PR will not make any change if the ipex version <= 2.3.0. 

Although the PR works well with our internal ipex version (2.3.0.dev), I will double-check it and make it compatible when the public ipex 2.3.0 is released.

I think it is ready to merge, would like to hear your opinion @ofirzaf . Thx!

echarlaix commented on tests/ipex/test_modeling.py:
```diff
@@ -255,6 +259,32 @@ def test_pipeline(self, model_arch):
+    def test_ipex_patching_generation(self, test_name, model_arch, use_cache):
+        model_id = MODEL_NAMES[model_arch]
+        set_seed(SEED)
+        model = IPEXModelForCausalLM.from_pretrained(model_id, export=True, use_cache=use_cache)
```
can you add 

```suggestion
        model = IPEXModelForCausalLM.from_pretrained(model_id, export=True, use_cache=use_cache)
        self.assertEqual(model.use_cache, use_cache)
```

jiqing-feng commented on tests/ipex/test_modeling.py:
```diff
@@ -128,7 +131,7 @@ def test_compare_to_transformers(self, model_arch):
         # Compare tensor outputs
         for output_name in {"logits", "last_hidden_state"}:
-            if output_name in transformers_outputs:
+            if output_name in transformers_outputs and output_name in outputs:
```
Actually `IPEXModel` didn't return `last_hidden_states` when `return_dict=False`. See [here](https://github.com/huggingface/optimum-intel/blob/main/optimum/intel/ipex/modeling_base.py#L196). `return_dict=True` is needed by jit trace.

echarlaix commented on optimum/intel/ipex/modeling_base.py:
```diff
@@ -414,7 +501,7 @@ def forward(
             position_ids.masked_fill_(attention_mask == 0, 1)
             if past_key_values:
-                position_ids = position_ids[:, -1].unsqueeze(-1)
+                position_ids = position_ids[:, -1].unsqueeze(0)
```
why this modification ? 
```suggestion
                position_ids = position_ids[:, -1].unsqueeze(-1)
```

echarlaix commented on tests/ipex/test_modeling.py:
```diff
@@ -259,6 +263,79 @@ def test_pipeline(self, model_arch):
+    @unittest.skipIf(is_ipex_version("<", "2.5.0"), reason="Only ipex version > 2.3.0 supports ipex model patching")
+    def test_ipex_patching_top_k(self, test_name, model_arch, use_cache):
+        model_id = MODEL_NAMES[model_arch]
+        set_seed(SEED)
+        model = IPEXModelForCausalLM.from_pretrained(model_id, export=True, use_cache=use_cache)
+        self.assertEqual(model.use_cache, use_cache)
+        tokenizer = AutoTokenizer.from_pretrained(model_id)
+        tokenizer.pad_token = tokenizer.eos_token
+        # Test with batch_size is 1 and 2.
+        texts = ["This is a sample", ["This is the first input", "This is the second input"]]
+        for text in texts:
+            tokens = tokenizer(text, padding=True, return_tensors="pt")
+            generation_config = GenerationConfig(max_new_tokens=4, do_sample=True, top_p=1.0, top_k=5)
+            outputs = model.generate(**tokens, generation_config=generation_config)
+            self.assertIsInstance(outputs, torch.Tensor)
+
+    @parameterized.expand(
+        grid_parameters(
+            {
+                "model_arch": IPEX_PATCHED_SUPPORTED_ARCHITECTURES,
+                "use_cache": [True, False],
+            }
+        )
+    )
+    @unittest.skipIf(is_ipex_version("<", "2.5.0"), reason="Only ipex version > 2.3.0 supports ipex model patching")
+    def test_ipex_patching_top_p(self, test_name, model_arch, use_cache):
+        model_id = MODEL_NAMES[model_arch]
+        set_seed(SEED)
+        model = IPEXModelForCausalLM.from_pretrained(model_id, export=True, use_cache=use_cache)
+        self.assertEqual(model.use_cache, use_cache)
+        tokenizer = AutoTokenizer.from_pretrained(model_id)
+        tokenizer.pad_token = tokenizer.eos_token
+        # Test with batch_size is 1 and 2.
+        texts = ["This is a sample", ["This is the first input", "This is the second input"]]
+        for text in texts:
+            tokens = tokenizer(text, padding=True, return_tensors="pt")
+            generation_config = GenerationConfig(max_new_tokens=4, do_sample=True, top_p=0.9, top_k=0)
+            outputs = model.generate(**tokens, generation_config=generation_config)
+            self.assertIsInstance(outputs, torch.Tensor)
+
```
```suggestion
    @parameterized.expand(
        grid_parameters(
            {
                "model_arch": IPEX_PATCHED_SUPPORTED_ARCHITECTURES,
                "use_cache": [True, False],
            }
        )
    )
    @unittest.skipIf(is_ipex_version("<", "2.5.0"), reason="Only ipex version > 2.3.0 supports ipex model patching")
    def test_ipex_patching_beam_search(self, test_name, model_arch, use_cache):
        model_id = MODEL_NAMES[model_arch]
        set_seed(SEED)
        model = IPEXModelForCausalLM.from_pretrained(model_id, export=True, use_cache=use_cache)
        self.assertEqual(model.use_cache, use_cache)
        trasnformers_model = AutoModelForCausalLM.from_pretrained(model_id)
        tokenizer = AutoTokenizer.from_pretrained(model_id)
        tokenizer.pad_token = tokenizer.eos_token
        # Test with batch_size is 1 and 2.
        texts = ["This is a sample", ["This is the first input", "This is the second input"]]
        generation_configs = (
            GenerationConfig(max_new_tokens=4, num_beams=2, do_sample=True),
            GenerationConfig(max_new_tokens=4, num_beams=4, do_sample=True),       
            GenerationConfig(max_new_tokens=4, num_beams=8, do_sample=True),       
            GenerationConfig(max_new_tokens=4, num_beams=32, do_sample=True),
            GenerationConfig(max_new_tokens=4, do_sample=not use_cache, top_p=1.0, top_k=5, penalty_alpha=0.6),
            GenerationConfig(max_new_tokens=4, do_sample=True, top_p=0.9, top_k=0),
       )
        for text in texts:
            tokens = tokenizer(text, padding=True, return_tensors="pt")
            for generation_config in generation_configs:
                outputs = model.generate(**tokens, generation_config=generation_config)
                transformers_outputs = trasnformers_model.generate(**tokens, generation_config=generation_config)
                self.assertIsInstance(outputs, torch.Tensor)
                self.assertEqual(outputs, transformers_outputs)
```

ofirzaf commented:
Isn't all this model patching suppose to be taken care by `ipex.llm.optimize`?

jiqing-feng commented:
> Isn't all this model patching suppose to be taken care by `ipex.llm.optimize`?

Hi @ofirzaf . The utility of ipex is support basic ops for the model just like pytorch. `ipex.llm.optimize` is just for our internal test and it does not compatible with the current transformer version.

jiqing-feng commented:
Hi @echarlaix . I have applied all your changes, thanks for that. The failed CI was caused by #589 , I think we could fix it soon. And this PR should be ready to merge : )
